### PR TITLE
typing: Improve documentation of generic classes and aliases

### PR DESCRIPTION
**Pull request by AlexWaygood**
The documentation for `typing.Dict` currently describes the alias as:

> A generic version of :class:`dict`.

In 2023, this is pretty misleading: it implies that `dict` is not generic, and that you have to use `typing.Dict` if you want to supply type parameters. There are similar issues across the typing documentation for the various aliases that were deprecated as part of PEP-585.

This PR also makes a few related edits to the docs for generic classes, which look like they still have some pre-PEP-585 language in them.

----
:books: Documentation preview :books:: https://cpython-previews--105369.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/typing.rst [hunk 1]
```diff
@@ -288,32 +288,39 @@ Generics
...
    from collections.abc import Sequence
    from typing import TypeVar
 
-   U = TypeVar('U')      # Declare type variable
+   U = TypeVar('U')                  # Declare type variable "U"
 
-   def first(l: Sequence[U]) -> U:   # Generic function
-       return l[0]
+   def second(l: Sequence[U]) -> U:  # Function is generic around the TypeVar "U"
+       return l[1]
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]: I figured it was slightly confusing to have two functions named `first` in two adjacent code examples

**Comment by AlexWaygood** [comment]
Docs preview: https://cpython-previews--105369.org.readthedocs.build/en/105369/library/typing.html#module-typing

File: Doc/library/typing.rst [hunk 2]
```diff
@@ -288,32 +288,39 @@ Generics
...
                        overrides: Mapping[str, str]) -> None: ...
 
-Generics can be parameterized by using :ref:`type parameter syntax <type-params>`::
+Generic functions and classes can be parameterized by using
+:ref:`type parameter syntax <type-params>`::
 
    from collections.abc import Sequence
 
-   def first[T](l: Sequence[T]) -> T:   # Generic function
+   def first[T](l: Sequence[T]) -> T:  # Function is generic around the TypeVar "T"
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]: Don't we usually say "generic over"?

File: Doc/library/typing.rst [hunk 3]
```diff
@@ -628,14 +635,14 @@ The module defines the following classes, functions and decorators.
 
 .. note::
 
-   This module defines several types that are subclasses of pre-existing
-   standard library classes which also extend :class:`Generic`
-   to support type variables inside ``[]``.
-   These types became redundant in Python 3.9 when the
+   This module defines several deprecated aliases to pre-existing
+   standard library classes. These were originally included in the typing
+   module in order to support type variables inside ``[]``.
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]: That seems inartfully worded: normally you put a concrete type (like `int`) inside the brackets, not a typevar.
Maybe "in order to support parameterizing these as generic classes"?

File: Doc/library/typing.rst [hunk 4]
```diff
@@ -909,12 +918,14 @@ These can be used as types in annotations using ``[]``, each having a unique syn
 
 .. data:: Callable
 
-   Callable type; ``Callable[[int], str]`` is a function of (int) -> str.
+   Deprecated alias to :class:`collections.abc.Callable`.
+
+   ``Callable[[int], str]`` signifies a function of (int) -> str.
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
   ``Callable[[int], str]`` signifies a function that takes a single parameter of type `int` and returns a `str`.
```
PEP 677 was rejected

File: Doc/library/typing.rst [hunk 5]
```diff
@@ -909,12 +918,14 @@ These can be used as types in annotations using ``[]``, each having a unique syn
...
 
    The subscription syntax must always be used with exactly two
    values: the argument list and the return type.  The argument list
-   must be a list of types or an ellipsis; the return type must be
-   a single type.
+   must be a list of types, a :class:`ParamSpec` or an ellipsis;
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]: Or Concatenate

File: Doc/library/typing.rst (hunk 4 above)
Comment (AlexWaygood) [reply Doc/library/typing.rst]: This is true, but the `a function of (int) -> str` language in the docs long predates PEP 677 -- it was introduced seven years ago, in https://github.com/python/cpython/commit/deed0c797ceb9e0cb2e8e9089ac717cbd2f7f9a8. I think it's fairly obvious what it means, without reference to PEP 677?

https://github.com/python/cpython/blob/27c68a6d8f20090310450862c2c299bb7ba3c160/Doc/library/typing.rst?plain=1#L939

The only change I'm making here is going from "is a function" to "signifies a function"

Comment (JelleZijlstra) [reply Doc/library/typing.rst]: Regardless of the history, I think it's better to write this out for clarity.

Comment (AlexWaygood) [reply Doc/library/typing.rst]: Okay, I'll make the same change to the docstring as well

Comment (AlexWaygood) [reply Doc/library/typing.rst]: Done in 1122321, which is annoyingly close to being a palindromic hash

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**